examples/ai/chat/mlx_chat.py

```diff
@@ -0,0 +1,206 @@
+# /// script
+# requires-python = ">=3.11"
+# dependencies = [
+#     "marimo",
+#     "mlx-lm",
+#     "huggingface_hub",
+# ]
+# ///
+
+import marimo
+
+__generated_with = "0.9.1"
+app = marimo.App(width="medium")
+
+
+@app.cell
+def __():
+    from mlx_lm import load, generate
+    from pathlib import Path
+    import marimo as mo
+    from huggingface_hub import snapshot_download
+    return mo, load, generate, Path, snapshot_download
+
+
+@app.cell
+def __(mo):
+    mo.md(
+        r"""
+        # Using MLX with Marimo
+        ## Chat Example
+        This example shows how to use [`mo.ui.chat`](https://docs.marimo.io/api/inputs/chat.html#marimo.ui.chat) to make a chatbot backed by Apple's MLX, using the `mlx_lm` library and marimo.
+        [`mlx_lm`](https://github.com/ml-explore/mlx-examples/tree/main/llm) is a library for running large language models on Apple Silicon.
+        [`mlx`](https://github.com/ml-explore/mlx) is a framework for running machine learning models on Apple Silicon.
+
+        Convert your own models to MLX, or find community-converted ones at various quantizations [here](https://huggingface.co/mlx-community).
+        
+        ### Things you can do to improve this example:
+        - [`prompt caching`](https://github.com/ml-explore/mlx-examples/blob/main/llms/README.md#long-prompts-and-generations)
+        - completions / notebook mode
+        - assistant pre-fill
+        """
+    )
+    return
+
+@app.cell
+def __(Path, snapshot_download):
+    def get_model_path(path_or_hf_repo: str) -> Path:
+        """
+        Ensures the model is available locally. If the path does not exist locally,
+        it is downloaded from the Hugging Face Hub.
+
+        Args:
+            path_or_hf_repo (str): The local path or Hugging Face repository ID of the model.
+
+        Returns:
+            Path: The path to the model.
+        """
+        model_path = Path(path_or_hf_repo)
+        if model_path.exists():
+            return model_path
+        else:
+            try:
+                # If it doesn't exist locally, download it from Hugging Face
+                return Path(snapshot_download(
+                    repo_id=path_or_hf_repo,
+                    allow_patterns=[
+                        "*.json",
+                        "*.safetensors",
+                        "*.py",
+                        "tokenizer.model",
+                        "*.tiktoken",
+                        "*.txt",
+                    ],
+                ))
+            except Exception as e:
+                raise ValueError(f"Error downloading model from Hugging Face: {str(e)}")
+
+    return (get_model_path,)
+
+@app.cell
+def __(mo):
+    MODEL_ID = mo.ui.text(
+        label="Hugging Face Model Repo or Local Path", 
+        value="mlx-community/Llama-3.2-3B-Instruct-bf16",
```

Comment:
just curious, is `mlx-community/Llama-3.2-3B-Instruct-bf16` the smallest model? i am finding it take a long time to download

Reply:
No there are smaller ones, this is just the  full precision one which is most accurate. You can get 8 bit or 4 bit quant, or also just get the smaller llama 3.2 1B parameter model. 

Here’s search results for all the post-converted MLX community llama 3.2 models. Note that instruct is the ‘chat’ one. Without it is the base model, which is autocomplete or finetuning your own use case. Base models are my favorite, and planning to do a “notebook” mode example next. 

https://huggingface.co/models?search=mlx%20llama%203.2